Add option to dump AST to files

vinivendra opened the pull request:
This adds two things:
- Calling `swiftc -dump-ast foo.swift [...] -o foo.ast` will dump the AST to the file `foo.ast`, instead of dumping it to `stderr` as usual.
- Calling `swiftc -dump-ast -output-file-map=outputFileMap.json *.swift [...]`, given an `outputFileMap.json` file that contains entries in the form `"ast-dump": "foo.ast"`, will dump the ASTs of the input files to their respective output files in the file map.

This should serve as a valid workaround to a bug mentioned in [the forums](https://forums.swift.org/t/error-when-dumping-the-ast-for-hundreds-of-files/17578) where the AST dump functionality crashes when called with too many input files. A few implementation details were also discussed in the same forum post.

As an aside, this also fixes a comment in `include/swift/Basic/PrimarySpecificPaths.h` that was incorrect.

vinivendra committed:
Add option to dump AST to files

This adds two things:
- Calling `swiftc -dump-ast foo.swift [...] -o foo.ast` will dump the AST to the file `foo.ast`, instead of dumping it to `stderr` as usual.
- Calling `swiftc -dump-ast -output-file-map=outputFileMap.json *.swift [...]`, given an `outputFileMap.json` file that contains entries in the form `"ast-dump": "foo.ast"`, will dump the ASTs of the input files to their respective output files in the file map.

This should serve as a valid workaround to a bug mentioned in [the forums](https://forums.swift.org/t/error-when-dumping-the-ast-for-hundreds-of-files/17578) where the AST dump functionality crashes when called with too many input files. A few implementation details were also discussed in the same forum post.

As an aside, this also fixes a comment in `include/swift/Basic/PrimarySpecificPaths.h` that was incorrect.

jrose-apple commented on lib/Basic/FileTypes.cpp:
```diff
@@ -147,6 +149,7 @@ bool file_types::isAfterLLVM(ID Id) {
 bool file_types::isPartOfSwiftCompilation(ID Id) {
   switch (Id) {
   case file_types::TY_Swift:
+  case file_types::TY_ASTDump:
```
This one's about inputs rather than outputs, so I'd prefer it go in the "false" section. It's not likely to come up, though.

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -1411,11 +1411,15 @@ void Driver::buildOutputInfo(const ToolChain &TC, const DerivedArgList &Args,
+
+    case options::OPT_dump_ast:
+      OI.CompilerOutputType = file_types::TY_ASTDump;
+      break;
```
Nitpick: can you put this next to the other "simple" cases, so it's clearer that they behave similarly?

vinivendra replied:
I think I put it there in part because it was next to (just before) the `Nothing` cases. Where would you rather I put it?

jrose-apple replied:
I was thinking up after `OPT_emit_bc`, since that's where the rest of the `-o`-ish options are.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -396,6 +396,8 @@ const char *ToolChain::JobContext::computeFrontendModeForCompile() const {
   case file_types::TY_PCH:
     return "-emit-pch";
+  case file_types::TY_ASTDump:
+      return "-dump-ast";
```
Stray indentation here.

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -685,6 +685,19 @@ static SourceFile *getPrimaryOrMainSourceFile(CompilerInvocation &Invocation,
 }
 
+static void dumpAST(CompilerInvocation &Invocation, CompilerInstance &Instance) {
+  SourceFile *sourceFile = getPrimaryOrMainSourceFile(Invocation, Instance);
```
Hm. This isn't going to work in batch mode, where you have more than one primary source file. Take a look at `emitLoadedModuleTraceForAllPrimariesIfNeeded` instead.

jrose-apple reviewed (commented):
Looking good, needs tests! I suggest adding some invocations to test/Driver/driver-compile.swift, and testing the frontend in a way similar to test/Frontend/batch-mode.swift.

vinivendra committed:
Add a few fixes mentioned in the code review.

vinivendra committed:
Fix the dumpAST function to work in batch mode.

Also removes the unnecessary `Invocation` parameter.

vinivendra committed:
Send `OPT_dump_ast` closer to other similar options.

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -685,6 +685,22 @@ static SourceFile *getPrimaryOrMainSourceFile(CompilerInvocation &Invocation,
+/// Dumps the AST of all available primary source files. If corresponding output
+/// files were specified, use them; otherwise, dump the AST to stderr.
+static void dumpAST(CompilerInstance &Instance) {
+  for(SourceFile *sourceFile: Instance.getPrimarySourceFiles()) {
```
Nitpick: We put a space between the `for` and the open paren.

jrose-apple replied:
Can you also add a FIXME comment that says this doesn't do the right thing for WMO? We don't actually have to fix that right now, since no one is asking for it, but it's good to call out deficiencies like that.

vinivendra replied:
What would the right thing be for WMO? You said there aren't any primary files, so could we check for that case and maybe dump all... "non-primary" files?

jrose-apple replied:
I think `forEachInputProducingAMainOutputFile` in FrontendInputsAndOutputs is probably closest to the right thing, since that's where you get output filenames from. It's *still* not quite going to be right for single-threaded WMO because you only get one output file in that case, which'll be repeatedly overwritten the way the code is written now, but that's even more marginal. Both Xcode and SwiftPM always build with multi-threaded WMO these days.

jrose-apple replied:
Actually, that doesn't give you SourceFiles. I guess when `getPrimarySourceFiles` is empty, you could iterate over all files from `getMainModule`. This doesn't handle single-threaded WMO either for the reason listed above, which you can special-case or not.

I do still think it's okay to ignore this whole mess for now because you don't need it.

jrose-apple replied:
Oh, another option that might be slightly more palatable than "ignore this whole mess" is "have the driver ignore WMO options when using -dump-ast, and warn about it". We have similar things where -wmo overrides -incremental.

vinivendra replied:
Hmmm I like this idea. But maybe for a future pull request.

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -685,6 +685,22 @@ static SourceFile *getPrimaryOrMainSourceFile(CompilerInvocation &Invocation,
+    if (OutputFilename == "-") {
+      sourceFile->dump();
+    } else {
+      auto OS = getFileOutputStream(OutputFilename, Instance.getMainModule()->getASTContext());
```
Nitpick: `Instance.getASTContext()`. Also, watch the 80-column limit!

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -685,6 +685,22 @@ static SourceFile *getPrimaryOrMainSourceFile(CompilerInvocation &Invocation,
+    const StringRef OutputFilename =
+      Instance.getPrimarySpecificPathsForSourceFile(*sourceFile).OutputFilename;
+    // PSP.OutputFilename is "-" whenever the output should be written to the console.
+    if (OutputFilename == "-") {
```
I think `getFileOutputStream` handles this, actually, by using `llvm::raw_fd_ostream`'s constructor. So you can drop this special case.

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -685,6 +685,22 @@ static SourceFile *getPrimaryOrMainSourceFile(CompilerInvocation &Invocation,
+  for(SourceFile *sourceFile: Instance.getPrimarySourceFiles()) {
+    const StringRef OutputFilename =
+      Instance.getPrimarySpecificPathsForSourceFile(*sourceFile).OutputFilename;
+    // PSP.OutputFilename is "-" whenever the output should be written to the console.
```
I think `getFileOutputStream` handles this, actually, by using `llvm::raw_fd_ostream`'s constructor. So you can use that.

vinivendra replied:
It seems that `llvm::raw_fd_ostream`'s constructor defaults to `stdout` when the filename is `"-"`, whereas the AST dump defaults to `stderr`.
I personally wouldn't mind changing the behavior to use `stdout` here, as it would slightly simplify both the code and my research... but maybe it's standard practice to use `stderr` for these things? I don't know.

vinivendra replied:
It's my understanding that we actually can't dump the ast at the same time as another output mode (which is why I have to remove the `-emit-...` options from the Xcode command), so yeah, I think it should be fine.

vinivendra replied:
It seems that dropping this makes the default constructor return a reference to `stdout` that isn't `llvm::outs()`. This causes the AST dumping code to dump everything without colors, since it checks for `&os == &llvm::outs()`, which fails. Do you know another way to check it the OS is `stdout`? Otherwise, I think I'm just gonna leave the manual check and default to `llvm::errs()` to preserve the colors.

jrose-apple replied:
Arguably that `has_colors` check is wrong. I don't see why we're not just asking the stream itself.

jrose-apple replied:
Did you want to do something about this for this PR, or follow up after?

vinivendra replied:
I fixed it in my last commit! I thought it’d show up here, maybe I’m doing git wrong heh sorry

jrose-apple replied:
Ah, but that check is copy/pasted all over ASTDumper…

vinivendra replied:
Oh, I hadn’t noticed... but yeah, I think I’d better do it in this PR. I’ll send an update tomorrow then.

vinivendra replied:
The other checks were setting `ShowColors` attributes in some (but not all) ast dump classes, but these attributes were never used again.
I removed the attributes (and the checks), so now we only do the check in one place, and it's done "correctly", and the classes are more consistent.

vinivendra replied:
Yeah, for the record I'm having trouble building master too, so I think this is OK. Let's see what the CI says.

jrose-apple reviewed (commented):
Some more feedback while you write tests, but this looks good!

vinivendra committed:
Fixes from the code review.

- Adds a space after the `for`;
- Changes `Instance` to use `getASTContext` directly.
- The `getFileOutputStream` method already checks for the `"-"` file name and returns `stdout` in that case. Replace the current (redundant) check with that one, and therefore replace the current default (`stderr`) with `stdout`.

vinivendra committed:
Add a FIXME to warn about potential WMO bugs.

vinivendra committed:
Add tests for frontend and driver.

jrose-apple commented on test/Driver/driver-compile.swift:
```diff
@@ -7,6 +7,10 @@
 // RUN: %FileCheck %s < %t.complex.txt
 // RUN: %FileCheck -check-prefix COMPLEX %s < %t.complex.txt
 
+// RUN: %swiftc_driver -driver-print-jobs -dump-ast -target x86_64-apple-macosx10.9 %s 2>&1 > %t.ast.txt
```
Can you also add a test with `-o`? That's a simple way of checking that we're treating this as a kind of main output.

jrose-apple commented on test/Frontend/ast-dump.swift:
```diff
@@ -0,0 +1,44 @@
+// MAIN-AST-NEXT: (tuple_expr type='()'
+// MAIN-AST: (call_expr
+// MAIN-AST-NEXT: decl=main.(file).b()@{{.*}}b.swift
+// MAIN-AST-NEXT: (tuple_expr type='()'
```
I suspect some of these details are going to change, like using a `tuple_expr` to represent function call arguments. Can you take those lines out, and maybe also the context chain part of the decl refs? (`decl={{.*}}b()@{{.*}}b.swift`, for example.)

jrose-apple reviewed (commented):
This looks good! I just have two comments for improvements to the tests, but then I think it's ready to go.

vinivendra committed:
Add a test for `-o` in the driver

vinivendra committed:
Simplify the ast-dump test to avoid hitting normal changes in the AST.

vinivendra committed:
Improve the color check in AST dump.

It should now work with the default `stdout` OS given to us by `getFileOutputStream` in `FrontendTools.cpp`.

vinivendra committed:
Remove other color checks from the AST dump that are never used.

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
Hm.

@swift-ci Please smoke test

jrose-apple commented:
Uh, hm. Looks like a bunch of tests were still expecting the AST dump output on stderr.

vinivendra commented:
> Uh, hm. Looks like a bunch of tests were still expecting the AST dump output on stderr.

Oh thanks, I couldn't find out why...
I think I'm gonna change the tests to expect `stdout` (rather than change the AST dump to `stderr`), what do you think?

vinivendra committed:
Fix broken tests

- Many tests got broken because of two things:
  - AST dump now outputs to stdout, but many tests expected stderr. This was a straightforward fix.
  - Many tests call swift with specific parameters; specifically, many call `swift frontend` directly. This makes them go through the compiler in unexpected ways, and specifically it makes them not have primary files, which breaks the new AST dump implementation. This commit adds the old implementation as a fallback for those cases, except it dumps to `stdout` to maintain some consistence.

Finally, the `/test/Driver/filelists.swift` failed for unknown reasons. It seems its output now had some lines out of order, and fixing the order made the test pass. However, as the reasons why it failed are unknown, this fix might not have been a good idea. Corrections are welcome.

vinivendra committed:
Fix an incorrect regex that matched more than it should

For example, the first regex matched the following text incorrectly (the `-filelist` argument matched until the first open double-quote in `"BUILD_DIR/\$(CONFIGURATION)...`).

````
BUILD_DIR/Debug/bin/swift -frontend -c -filelist /var/folders/.../sources-098836 -primary-filelist /var/folders/.../primaryInputs-b8aa5a -supplementary-output-file-map /var/folders/.../supplementaryOutputs-b1a037 -target x86_64-apple-macosx10.9 -enable-objc-interop -module-cache-path "BUILD_DIR/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache" -swift-version 4 -module-name main -output-filelist /var/folders/.../outputs-c97286
````

Also removes unused variables.

vinivendra committed:
Fix attr_objc test to check AST dump in `stdout`

Also change the file extensions to `.ast` just to match those in other files.

jrose-apple commented:
@swift-ci Please test
